### PR TITLE
only consider edges reused if not a superset path

### PR DESCRIPTION
**Pull request by jhsinger-klotho**
making sure edges arent part of a superset path when determining if they should affect edge deletion in path removal

testing repro example
```
constraints:
- operator: must_not_exist
  scope: edge
  target:
    source: aws:api_integration:rest_api_0:rest_api_0_integration_0
    target: aws:lambda_function:lambda_function_2
resources:
    aws:api_stage:rest_api_0:api_stage-0:
        Deployment: aws:api_deployment:rest_api_0:api_deployment-0
        RestApi: aws:rest_api:rest_api_0
        StageName: stage
    aws:api_deployment:rest_api_0:api_deployment-0:
        RestApi: aws:rest_api:rest_api_0
        Triggers:
            rest_api_0_integration_0: rest_api_0_integration_0
            rest_api_0_integration_0_method: rest_api_0_integration_0_method
    aws:rest_api:rest_api_0:
        BinaryMediaTypes:
            - application/octet-stream
            - image/*
    aws:api_resource:rest_api_0:api_resource-0:
        FullPath: /{proxy+}
        PathPart: '{proxy+}'
        RestApi: aws:rest_api:rest_api_0
    aws:api_method:rest_api_0:rest_api_0_integration_0_method:
        Authorization: NONE
        HttpMethod: ANY
        RequestParameters:
            method.request.path.proxy: true
        Resource: aws:api_resource:rest_api_0:api_resource-0
        RestApi: aws:rest_api:rest_api_0
    aws:api_integration:rest_api_0:rest_api_0_integration_0:
        IntegrationHttpMethod: POST
        Method: aws:api_method:rest_api_0:rest_api_0_integration_0_method
        RequestParameters:
            integration.request.path.proxy: method.request.path.proxy
        Resource: aws:api_resource:rest_api_0:api_resource-0
        RestApi: aws:rest_api:rest_api_0
        Route: /{proxy+}
        Target: aws:lambda_function:lambda_function_2
        Type: AWS_PROXY
        Uri: aws:lambda_function:lambda_function_2#LambdaIntegrationUri
    aws:lambda_permission:rest_api_0_integration_0-lambda_function_2:
        Action: lambda:InvokeFunction
        Function: aws:lambda_function:lambda_function_2
        Principal: apigateway.amazonaws.com
        Source: aws:rest_api:rest_api_0#ChildResources
    aws:lambda_function:lambda_function_2:
        ExecutionRole: aws:iam_role:lambda_function_2-ExecutionRole
        Image: aws:ecr_image:lambda_function_2-image
        LogGroup: aws:log_group:lambda_function_2-log-group
        MemorySize: 512
        Timeout: 180
    aws:SERVICE_API:lambda_function_2-lambda_function_2-log-group:
    aws:ecr_image:lambda_function_2-image:
        Context: .
        Dockerfile: lambda_function_2-image.Dockerfile
        Repo: aws:ecr_repo:ecr_repo-0
    aws:iam_role:lambda_function_2-ExecutionRole:
        AssumeRolePolicyDoc:
            Statement:
                - Action:
                    - sts:AssumeRole
                  Effect: Allow
                  Principal:
                    Service:
                        - lambda.amazonaws.com
            Version: "2012-10-17"
        ManagedPolicies:
            - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
    aws:ecr_repo:ecr_repo-0:
        ForceDelete: true
    aws:log_group:lambda_function_2-log-group:
        LogGroupName: /aws/lambda/lambda_function_2
        RetentionInDays: 5
edges:
    aws:api_stage:rest_api_0:api_stage-0 -> aws:api_deployment:rest_api_0:api_deployment-0:
    aws:api_stage:rest_api_0:api_stage-0 -> aws:rest_api:rest_api_0:
    aws:api_deployment:rest_api_0:api_deployment-0 -> aws:api_integration:rest_api_0:rest_api_0_integration_0:
    aws:api_deployment:rest_api_0:api_deployment-0 -> aws:api_method:rest_api_0:rest_api_0_integration_0_method:
    aws:api_deployment:rest_api_0:api_deployment-0 -> aws:rest_api:rest_api_0:
    aws:rest_api:rest_api_0 -> aws:api_integration:rest_api_0:rest_api_0_integration_0:
    aws:rest_api:rest_api_0 -> aws:api_method:rest_api_0:rest_api_0_integration_0_method:
    aws:rest_api:rest_api_0 -> aws:api_resource:rest_api_0:api_resource-0:
    aws:api_resource:rest_api_0:api_resource-0 -> aws:api_integration:rest_api_0:rest_api_0_integration_0:
    aws:api_resource:rest_api_0:api_resource-0 -> aws:api_method:rest_api_0:rest_api_0_integration_0_method:
    aws:api_method:rest_api_0:rest_api_0_integration_0_method -> aws:api_integration:rest_api_0:rest_api_0_integration_0:
    aws:api_integration:rest_api_0:rest_api_0_integration_0 -> aws:lambda_permission:rest_api_0_integration_0-lambda_function_2:
    aws:lambda_permission:rest_api_0_integration_0-lambda_function_2 -> aws:lambda_function:lambda_function_2:
    aws:lambda_function:lambda_function_2 -> aws:SERVICE_API:lambda_function_2-lambda_function_2-log-group:
    aws:lambda_function:lambda_function_2 -> aws:ecr_image:lambda_function_2-image:
    aws:lambda_function:lambda_function_2 -> aws:iam_role:lambda_function_2-ExecutionRole:
    aws:SERVICE_API:lambda_function_2-lambda_function_2-log-group -> aws:log_group:lambda_function_2-log-group:
    aws:ecr_image:lambda_function_2-image -> aws:ecr_repo:ecr_repo-0:
    aws:iam_role:lambda_function_2-ExecutionRole -> aws:log_group:lambda_function_2-log-group:

```

### Standard checks

- **Unit tests**: Any special considerations?
- **Docs**: Do we need to update any docs, internal or public?
- **Backwards compatibility**: Will this break existing apps? If so, what would be the extra work required to keep them working?

### CODE REVIEW

File: pkg/engine2/reconciler/remove_edge_test.go [hunk 1]
```diff
@@ -139,6 +139,46 @@ func Test_findEdgesUsedInOtherPathSelection(t *testing.T) {
 				{Source: construct.ResourceId{Provider: "p", Type: "t", Name: "t"}, Target: construct.ResourceId{Provider: "p", Type: "t", Name: "b"}},
 			},
 		},
+		{
+			name:         "path is superset of deletion path",
+			source:       construct.ResourceId{Provider: "p", Type: "t", Name: "s"},
+			target:       construct.ResourceId{Provider: "p", Type: "t", Name: "b"},
+			initialState: []any{"p:t:s", "p:t:t", "p:t:t -> p:t:s", "p:t:a", "p:t:b", "p:t:a -> p:t:t", "p:t:s -> p:t:b", "p:t:l", "p:t:l -> p:t:b"},
```
Comment (gordon-klotho) [inline pkg/engine2/reconciler/remove_edge_test.go:146]: nit: the single nodes aren't necessary unless they don't have edges, and you can specify full paths if you want
```
[]any{"p:t:a -> p:t:t -> p:t:s -> p:t:b", "p:t:l -> p:t:b"}
```

File: pkg/engine2/reconciler/remove_edge.go [hunk 2]
```diff
@@ -47,7 +47,10 @@ func RemovePath(
 	// We will pass the explicit field as false so that explicitly added resources do not get deleted
 	for _, path := range paths {
 		for _, resource := range path {
-			errs = errors.Join(errs, RemoveResource(ctx, resource, false))
+			err := RemoveResource(ctx, resource, false)
+			if err != nil && !errors.Is(err, graph.ErrVertexNotFound) {
```
Comment (gordon-klotho) [inline pkg/engine2/reconciler/remove_edge.go:51]: When could this return `ErrVertexNotFound`? That'd be a surprising result without knowing more. Can you add a comment how that could happen and why it's okay to ignore?

**Review by gordon-klotho — COMMENTED** [review]
Can you put the repro input as a functional test case?

**Comment by jhsinger-klotho** [comment]
> Can you put the repro input as a functional test case?

sure, but i think at some point the functional tests will not become maintainable if we add each specific use case to them, rather than just relying on unit tests, which i added to catch this case

**Comment by gordon-klotho** [comment]
> at some point the functional tests will not become maintainable if we add each specific use case to them, rather than just relying on unit tests, which i added to catch this case

I agree, but I don't think we have an edge removal case do we? Just a resource removal but maybe I'm mistaken

**Review by gordon-klotho — APPROVED** [review]

**Pull request merged**